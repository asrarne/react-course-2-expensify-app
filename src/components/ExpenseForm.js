import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
// import Decimal from 'decimal.js';


export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }
    

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onSubmit = (event) => {
        event.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({error: 'Please provide description and amount.'}));
        } else {
            this.setState(()=> ({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                note: this.state.note,
                //(Math.round(num1 * 100) / 100).toFixed(2))
                //(Math.round(this.state.amount * 100)).toFixed(2))
                amount: parseFloat((parseFloat(this.state.amount, 10) * 100.00).toFixed(2)),
                // amount: (Math.round(this.state.amount * 100)/100).toFixed(2),
                createdAt: this.state.createdAt.valueOf() 
            });
            // console.log('submitted', this.state.amount);
            // console.log('submitted', parseFloat(this.state.amount).toFixed(2) * 100);
            // console.log('submitted', typeof((parseFloat(this.state.amount, 10) * 100).toFixed(2)));
            // console.log('submitted', typeof(parseFloat((parseFloat(this.state.amount, 10) * 100.00).toFixed(2))));
            // console.log('submitted', typeof(parseFloat(this.state.amount, 10) * 100.00));
        }
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(()=>({ amount }));
            // console.log('amount: ', amount);
            // console.log('this.state.amount: ', this.state.amount);
            // console.log('parseFloat: ', parseFloat(this.state.amount, 10) * 100 );
        }
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({calendarFocused: focused}));
    };

    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input 
                    type="text" 
                    placeholder="Description"
                    className="text-input" 
                    autoFocus 
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                    />
                <input 
                    type="text" 
                    placeholder="Amount"
                    className="text-input"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                    />
                <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />    
                <textarea 
                    placeholder="Add a note for your expense (optional)"
                    className="textarea"
                    value = {this.state.note}
                    onChange={this.onNoteChange}
                    ></textarea>
                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        )
    }
}