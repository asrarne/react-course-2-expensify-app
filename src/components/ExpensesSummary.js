import React from "react";
import { connect } from "react-redux";
import selectExpenseTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = (props) => (
    <div>
        <p>Viewing {props.expenseCount} { props.expenseCount > 1 ? 'expenses' : 'expense' } totalling {numeral(props.expensesTotal/100).format('$0,0.00')}</p>
    </div>
);

const mapStateToProps = (state) => ({
    expenseCount: state.expenses.length,
    expensesTotal: selectExpenseTotal(state.expenses)
});

export default connect(mapStateToProps)(ExpensesSummary);