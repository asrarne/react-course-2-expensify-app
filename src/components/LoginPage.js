import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';
import {
        googleAuthProvider,
        facebookAuthProvider,
        twitterAuthProvider,
        githubAuthProvider
        } from '../firebase/firebase';

export class LoginPage extends React.Component {
    onFaceBookClick = () => {
        this.props.startLogin(this.props.facebookAuthProvider);
    };

    onGoogleClick = () => {
        this.props.startLogin(this.props.googleAuthProvider);
    };

    onTwitterClick = () => {
        this.props.startLogin(this.props.twitterAuthProvider);
    };

    onGithubClick = () => {
        this.props.startLogin(this.props.githubAuthProvider);
    };

    render () {
        return (
            <div className="box-layout">
                <div className="box-layout__box">
                    <h1 className="box-layout__title">Expensify</h1>
                    <p>It's time to get your expenses under control.</p>
                    <div className="box-layout__action">
                        <button className="button" onClick={this.onGoogleClick}>Login with Google</button>
                        <button className="button" onClick={this.onFaceBookClick}>Login with Facebook</button>
                        <button className="button" onClick={this.onTwitterClick}>Login with Twitter</button>
                        <button className="button" onClick={this.onGithubClick}>Login with GitHub</button>
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = () => ({
    googleAuthProvider: googleAuthProvider,
    facebookAuthProvider: facebookAuthProvider,
    twitterAuthProvider: twitterAuthProvider,
    githubAuthProvider: githubAuthProvider
});

const mapDispatchTOProps = (dispatch) => ({
    startLogin: (data) => dispatch(startLogin(data))
});

export default connect(mapStateToProps, mapDispatchTOProps)(LoginPage);