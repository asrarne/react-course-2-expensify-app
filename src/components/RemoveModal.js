import React from 'react';
import Modal from 'react-modal';


const RemoveModal = (props) => (
    <Modal
        isOpen={props.removeClicked}
        onRequestClose={props.handleRemoveNo}
        ariaHideApp={false}
        contentLabel="Do you want to remove expense?"
        closeTimeoutMS={200}
        className="modal"
    >
        <h3 className="modal__title"> Do you want to remove expense?</h3>
        <div className="modal__action">
        <button className="button" onClick={props.handleRemoveYes}>Yes</button>
        <button className="button" onClick={props.handleRemoveNo}>No</button>
        </div>
    </Modal>
);

export default RemoveModal;