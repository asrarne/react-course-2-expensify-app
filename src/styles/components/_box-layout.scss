.box-layout {
    background: url('/images/bg.jpg');
    background-size: cover;
    height: 100vh;
    width: 100vw;
    display: flex;
    align-items: center;
    justify-content: center;
}

.box-layout__box {
    background: fade-out(white, .15);
    border-radius: 3px;
    padding: $l-size $m-size;
    text-align: center;
    width: 30rem;
}

.box-layout__title {
    margin: 0 0 $m-size 0;
    line-height: 1;
}

.box-layout__action {
    // align-items: center;
    display: flex;
    flex-direction: column;
    // justify-content: space-between;
    // padding: $s-size;
    > * {
        margin-bottom: $m-size;
    }

    // margin-top: $s-size;
    // margin-bottom: $s-size;
}