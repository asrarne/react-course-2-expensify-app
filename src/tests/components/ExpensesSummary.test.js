import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should render ExpensesSummary with 1 expense and no hidden expense without filter, correctly', () => {
    const wrapper = shallow(<ExpensesSummary 
                                expenseCount={1} 
                                expensesTotal={235}
                                hiddenExpenseCount={0}
                                noFilterInPlace={true}
                                />);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with 1 expense and no hidden expense with filters, correctly', () => {
    const wrapper = shallow(<ExpensesSummary 
                                expenseCount={1} 
                                expensesTotal={235}
                                hiddenExpenseCount={0}
                                noFilterInPlace={false}
                                />);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with 1 expense and 1 hidden expense with filters, correctly', () => {
    const wrapper = shallow(<ExpensesSummary 
                                expenseCount={1} 
                                expensesTotal={235}
                                hiddenExpenseCount={0}
                                noFilterInPlace={false}
                                />);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with 1 expense and multiple hidden expense with filters, correctly', () => {
    const wrapper = shallow(<ExpensesSummary 
                                expenseCount={1} 
                                expensesTotal={235}
                                hiddenExpenseCount={2}
                                noFilterInPlace={false}
                                />);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with multiple expenses and no hidden expense without filters, correctly', () => {
    const wrapper = shallow(<ExpensesSummary 
                                expenseCount={23} 
                                expensesTotal={23456789}
                                hiddenExpenseCount={0}
                                noFilterInPlace={true}
                            />);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with multiple expenses and no hidden expense with filters, correctly', () => {
    const wrapper = shallow(<ExpensesSummary 
                                expenseCount={23} 
                                expensesTotal={23456789}
                                hiddenExpenseCount={0}
                                noFilterInPlace={false}
                            />);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with multiple expenses and 1 hidden expense with filters, correctly', () => {
    const wrapper = shallow(<ExpensesSummary 
                                expenseCount={23} 
                                expensesTotal={23456789}
                                hiddenExpenseCount={1}
                                noFilterInPlace={false}
                            />);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with multiple expenses and multiple hidden expenses with filters, correctly', () => {
    const wrapper = shallow(<ExpensesSummary 
                                expenseCount={23} 
                                expensesTotal={23456789}
                                hiddenExpenseCount={10}
                                noFilterInPlace={false}
                            />);
    expect(wrapper).toMatchSnapshot();
});