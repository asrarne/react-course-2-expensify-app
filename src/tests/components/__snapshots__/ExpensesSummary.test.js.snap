// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with 1 expense and 1 hidden expense with filters, correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $2.35
      </span>
    </h1>
    <p
      className="page-header__title"
    >
      Showing all expenses
    </p>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with 1 expense and multiple hidden expense with filters, correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $2.35
      </span>
    </h1>
    <p
      className="page-header__title"
    >
      Not showing 2 expenses because of filter
    </p>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with 1 expense and no hidden expense with filters, correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $2.35
      </span>
    </h1>
    <p
      className="page-header__title"
    >
      Showing all expenses
    </p>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with 1 expense and no hidden expense without filter, correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        $2.35
      </span>
    </h1>
    <p
      className="page-header__title"
    >
      Showing all expenses, No filter in place
    </p>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with multiple expenses and 1 hidden expense with filters, correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        23
      </span>
       
      expenses
       totalling 
      <span>
        $234,567.89
      </span>
    </h1>
    <p
      className="page-header__title"
    >
      Not showing 1 expense because of filter
    </p>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with multiple expenses and multiple hidden expenses with filters, correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        23
      </span>
       
      expenses
       totalling 
      <span>
        $234,567.89
      </span>
    </h1>
    <p
      className="page-header__title"
    >
      Not showing 10 expenses because of filter
    </p>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with multiple expenses and no hidden expense with filters, correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        23
      </span>
       
      expenses
       totalling 
      <span>
        $234,567.89
      </span>
    </h1>
    <p
      className="page-header__title"
    >
      Showing all expenses
    </p>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpensesSummary with multiple expenses and no hidden expense without filters, correctly 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        23
      </span>
       
      expenses
       totalling 
      <span>
        $234,567.89
      </span>
    </h1>
    <p
      className="page-header__title"
    >
      Showing all expenses, No filter in place
    </p>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
